### ISSUE TITLE
Implement AST-Based Answer Generation (Reasoning Core → Response)

### ISSUE THREAD

**Issue by marctjones** [marctjones/klareco#36]
## Overview

Implement the final piece that converts reasoning core output (AST representation) into natural language responses. This connects the reasoning core to the deparser for end-to-end abstractive QA.

### What We Have

**Input side (Query → Reasoning)**:
- ✅ Parser: Text → AST
- ✅ Compositional Embeddings: AST → vectors
- ✅ Tree-LSTM: AST → 256d representation
- ✅ RAG: Retrieve relevant facts

**Reasoning**:
- 🚧 Reasoning Core (Issue #23): Query AST + Retrieved ASTs → Answer representation

**Output side**:
- ✅ Deparser: AST → Text

### ❌ What's Missing

**The connection between reasoning core output and deparser**!

```python
# Reasoning core produces:
answer_repr = reasoning_core(query_ast, retrieved_asts)  # 512d vector

# But deparser needs:
answer_ast = ???  # AST structure

# Then:
answer_text = deparse(answer_ast)  # Final response
```

**Gap**: How do we convert `answer_repr` (vector) → `answer_ast` (structure)?

### The AST-as-Thoughts Philosophy

From `AST-as-Thoughts-Architecture.md`:

> "The reasoning core doesn't generate tokens - it generates AST structures that encode grammatical and semantic intent."

**We need**:
- Reasoning core output → AST structure
- NOT: Reasoning core output → text tokens (that's what traditional LLMs do)

### Two Approaches

**Approach A: Direct AST Decoding** (Recommended for M1)
```python
class ASTDecoder(nn.Module):
    """
    Decode vector representation to AST structure.
    
    Strategy: Predict AST nodes sequentially using autoregressive decoding.
    """
    
    def forward(self, answer_repr: Tensor) -> Dict:
        """
        answer_repr: [512d] from reasoning core
        
        Returns: AST dict structure
        {
            "tipo": "frazo",
            "subjekto": {...},
            "verbo": {...},
            "objekto": {...}
        }
        """
        # Decode subject
        subj_repr = self.subj_decoder(answer_repr)
        subjekto = self.decode_phrase(subj_repr)
        
        # Decode verb (conditioned on subject)
        verb_repr = self.verb_decoder(torch.cat([answer_repr, subj_repr], dim=-1))
        verbo = self.decode_word(verb_repr)
        
        # Decode object (conditioned on subject + verb)
        obj_repr = self.obj_decoder(torch.cat([answer_repr, subj_repr, verb_repr], dim=-1))
        objekto = self.decode_phrase(obj_repr)
        
        return {
            "tipo": "frazo",
            "subjekto": subjekto,
            "verbo": verbo,
            "objekto": objekto
        }
```

**Approach B: Template-Based Assembly** (Fallback for M1)
```python
class TemplateASTAssembler:
    """
    Use retrieval + reasoning to fill AST templates.
    
    Strategy: 
    1. Classify question type (Kiu/Kio/Kie/etc)
    2. Select answer template
    3. Fill slots from retrieved facts
    4. Use reasoning core to select best facts
    """
    
    TEMPLATES = {
        "identity": {  # "Kiu estas X?" → "X estas Y."
            "tipo": "frazo",
            "subjekto": "{entity}",
            "verbo": {"radiko": "est", "tempo": "prezenco"},
            "objekto": "{description}"
        },
        "definition": {  # "Kio estas X?" → "X estas Y."
            # Same structure
        },
        "location": {  # "Kie estas X?" → "X estas en Y."
            "tipo": "frazo",
            "subjekto": "{entity}",
            "verbo": {"radiko": "est", "tempo": "prezenco"},
            "objekto": {
                "tipo": "vortgrupo",
                "prepozicio": "en",
                "kerno": "{location}"
            }
        }
    }
```

### Option 1: Hybrid Approach (Recommended)

**Use templates for M1, train decoder for M2**:

1. **M1 (Simple)**:
   - Question classifier: Identify question type from query AST
   - Template selector: Pick answer template
   - Slot filler: Use reasoning core to select best facts for slots
   - AST assembler: Fill template → complete AST
   - Deparser: AST → text

2. **M2 (Learned)**:
   - Train full AST decoder on M1 data
   - Replace template system with learned generation
   - Keep templates as fallback

**Advantages**:
- ✅ M1: Fast, deterministic, debuggable
- ✅ M2: Flexible, handles complex answers
- ✅ Gradual complexity increase

#### Phase 1: Template System (M1 - 2 weeks)

**Create**: `klareco/answer_generation/template_assembler.py`

```python
class QuestionClassifier:
    """Classify question type from AST."""
    
    QUESTION_WORDS = {
        "kiu": "identity",    # Who
        "kio": "definition",  # What
        "kie": "location",    # Where
        "kiam": "time",       # When
        "kial": "reason",     # Why
        "kiel": "manner",     # How
        "kiom": "quantity",   # How much/many
    }
    
    def classify(self, query_ast: Dict) -> str:
        """Extract question word from AST."""
        # Find interrogative pronoun
        # Return question type
        pass

class TemplateASTAssembler:
    """Fill AST templates with facts from reasoning core."""
    
    def __init__(self, reasoning_core, templates):
        self.reasoning_core = reasoning_core
        self.templates = templates
        self.classifier = QuestionClassifier()
    
    def generate_answer_ast(
        self,
        query_ast: Dict,
        retrieved_facts: List[Dict]
    ) -> Dict:
        """
        Generate answer AST using templates.
        
        Flow:
        1. Classify question type
        2. Get template for that type
        3. Use reasoning core to select facts for slots
        4. Fill template → answer AST
        """
        # Classify
        q_type = self.classifier.classify(query_ast)
        template = self.templates[q_type]
        
        # Get slots to fill
        slots = self.extract_slots(template)
        
        # Use reasoning core to rank facts
        fact_scores = self.reasoning_core.score_facts(
            query_ast, retrieved_facts
        )
        
        # Fill slots from top-ranked facts
        filled_ast = self.fill_template(
            template, slots, fact_scores, retrieved_facts
        )
        
        return filled_ast
```

**Templates** (20-30 question types):
```python
ANSWER_TEMPLATES = {
    "identity": {
        "pattern": "X estas Y",
        "ast": {
            "tipo": "frazo",
            "subjekto": "{entity}",
            "verbo": {"radiko": "est", "tempo": "prezenco"},
            "objekto": "{description}"
        }
    },
    # ... 19-29 more templates
}
```

#### Phase 2: Learned Decoder (M2 - 3-4 weeks)

**Create**: `klareco/models/ast_decoder.py`

```python
class ASTDecoder(nn.Module):
    """
    Learned decoder: vector → AST structure.
    
    Architecture:
    - Autoregressive decoding
    - Predicts AST nodes one at a time
    - Conditioned on previous nodes (like language model)
    """
    
    def __init__(self, input_dim: int = 512, hidden_dim: int = 256):
        super().__init__()
        
        # Node type classifier
        self.node_type_head = nn.Linear(hidden_dim, len(NODE_TYPES))
        
        # Root predictor
        self.root_head = nn.Linear(hidden_dim, len(ROOT_VOCAB))
        
        # Grammar predictor
        self.grammar_head = nn.Linear(hidden_dim, len(GRAMMAR_FEATURES))
        
        # Decoder RNN
        self.decoder_rnn = nn.LSTM(input_dim, hidden_dim, num_layers=2)
    
    def forward(self, answer_repr: Tensor) -> Dict:
        """Decode vector to AST structure."""
        # Autoregressive decoding
        nodes = []
        hidden = self.init_hidden(answer_repr)
        
        while not self.is_complete(nodes):
            # Predict next node
            node_type = self.predict_node_type(hidden)
            node_content = self.predict_node_content(hidden, node_type)
            nodes.append((node_type, node_content))
            
            # Update hidden state
            hidden = self.step(hidden, node_type, node_content)
        
        # Assemble nodes → AST
        ast = self.assemble_ast(nodes)
        return ast
```

**Training**:
- Train on Q&A pairs from Issue #24
- Loss: Match predicted AST to target AST
- Metric: AST exact match, partial match (by role)

### M1 (Template System)

**Week 1**:
- [ ] Design 20-30 question templates
- [ ] Implement QuestionClassifier
- [ ] Implement TemplateASTAssembler

**Week 2**:
- [ ] Integrate with reasoning core (#25)
- [ ] Test on 100 Q&A examples
- [ ] Evaluate template coverage (% questions handled)

### M2 (Learned Decoder)

**Later** (after M1 complete):
- [ ] Implement ASTDecoder architecture
- [ ] Train on M1 Q&A dataset
- [ ] Evaluate vs template system
- [ ] Deploy if better, else keep templates

## Dependencies

**Blocks**:
- Issue #23: Reasoning Core Architecture (needs this for output)
- Issue #25: Training Reasoning Core (full end-to-end pipeline)

**Blocked by**:
- Issue #24: Q&A Training Data (examples for templates)

**Can work in parallel with**:
- Issue #28: Corpus building
- Issue #34: Embedding retraining

## Success Criteria

✅ **M1 Template System**:
- Covers 80%+ of factual questions
- Generates grammatically correct ASTs (100%)
- Integrates with reasoning core
- Deparser produces natural Esperanto

✅ **M2 Learned Decoder** (Future):
- Matches template accuracy on covered questions
- Handles edge cases templates miss
- Generalizes to unseen question types

## Priority

**P0** - Critical Path

This is required to complete the M1 pipeline:
1. Issue #23: Design reasoning core
2. **THIS ISSUE**: Design answer generation
3. Issue #25: Train end-to-end

## Related Issues

- Issue #23: Reasoning Core Architecture
- Issue #24: Q&A Training Data (provides template examples)
- Issue #25: Train Reasoning Core (needs this component)
- Issue #35: RAG Integration (extractive baseline to compare)

## Open Questions

1. **Should we implement both template + learned decoder for M1?**
   - Pro: Flexibility, better results
   - Con: More complexity, longer development

2. **How many templates do we need for M1?**
   - Minimum: 10 (core question types)
   - Target: 20-30 (comprehensive coverage)
   - Maximum: 50+ (diminishing returns)

3. **Can templates handle multi-sentence answers?**
   - M1: Single sentence only
   - M2: Multi-sentence with learned decoder
